src/core/ObservableQuery.ts

```diff
@@ -303,8 +303,15 @@ export class ObservableQuery<
 
   // Compares newResult to the snapshot we took of this.lastResult when it was
   // first received.
-  public isDifferentFromLastResult(newResult: ApolloQueryResult<TData>) {
-    return !this.last || !equal(this.last.result, newResult);
+  public isDifferentFromLastResult(
+    newResult: ApolloQueryResult<TData>,
+    variables?: TVariables
+  ) {
+    return (
+      !this.last ||
+      !equal(this.last.result, newResult) ||
+      !equal(this.last.variables, variables)
```

Comment:
```suggestion
      (variables && !equal(this.last.variables, variables))
```
In case `isDifferentFromLastResult` is called without `variables`, I think we should ignore the equality test?

Reply:
Hmm - couldn't there be a case in which variables go from defined to undefined and we'd still want to iterate if the data passes the equality check?

Reply:
I think we need to preserve the current behavior of `isDifferentFromLastResult` for callers that only pass `newResult` and no `variables`. Calling `isDifferentFromLastResult` directly is probably not common, but it's a public method.

When `variables` is falsy here, that usually means `isDifferentFromLastResult` was called without `variables`. In those cases (e.g. any current usage), I'm suggesting we should skip checking `equal(this.last.variables, variables)`, because it's likely to be false (just because `variables` is undefined), and also because the caller of `isDifferentFromLastResult` is asking only about the result, as indicated by passing no `variables`.

This logic isn't completely sound, because the `variables` argument could be passed explicitly as a falsy value (for example, becoming undefined). We could check `arguments.length` to determine if one or two arguments were passed to `isDifferentFromLastResult`, to rule out that possibility. Does that seem worthwhile?

Reply:
Got it, yep sounds good!

I just confirmed the `variables` arg passed to `Observer.next` inside of `reobserve` [defaults to an empty object](https://github.com/apollographql/apollo-client/blob/94256e6b715bc4cdaa3b3200e142caec4667075c/src/core/ObservableQuery.ts#L837), so the point I raised is moot/no upside in checking arguments length.